lib/Transforms/LiquidFixpointBuilder/VariablesEnvironmentImmutable.cpp

```diff
@@ -0,0 +1,577 @@
+    
+    {
+      auto createdBinderRes = constraintBuilder.CreateBinderWithConstraints(transitionGuardName, FixpointType::GetBoolType(),  { assignedExpr });
+      if (!createdBinderRes.Succeeded) { return createdBinderRes; }
+    }
+
+    return ResultType::Success();
+  }
+
+  ResultType VariablesEnvironmentImmutable::validateFinishedAndUnfinishedBlockStates(
+    const std::vector<std::string>& previousFinishedBlocks,
+    const std::vector<std::string>& previousUnfinishedBlocks)
+  {
+    //@TODO::(juspreet) - The logic is odd. It might be the cause of problem in loops. Verify.
+    if (previousFinishedBlocks.size() == 0 || previousFinishedBlocks.size() > 1) {
+      return ResultType::Error("Blocks were not completed in the right order. Expected one Completed Block, got "s + std::to_string(previousFinishedBlocks.size()));
+    }
+
+    // Ensure that :
+    // 1) forall blocks in PreviousFinishedBlocks, dominate(blocks, block), forall block in PreviousUnfinishedblocks
+    for (const auto& previousUnfinishedBlock : previousFinishedBlocks)
+    {
+      bool allDominated = true;
+      for (const auto& previousFinishedBlock : previousUnfinishedBlocks)
+      {
+	bool dominated;
+	{
+	  auto domRes = functionBlockGraph.StrictlyDominates(previousFinishedBlock, previousUnfinishedBlock, dominated);
+	  if (!domRes.Succeeded) { return domRes; }
+	}
+
+	if (!dominated)
+	{
+	  allDominated = false;
+	  break;
+	}
+      }
+
+      // If all blocks are not dominated, then exit with an appropriate error message
+      if (!allDominated)
+      {
+	return ResultType::Error("Blocks were not completed in the right order. Unexpected control flow graph. Finished: "s
+			    + RefinementUtils::StringJoin(", "s, previousFinishedBlocks)
+			    + " . Unfinished: "s
+			    + RefinementUtils::StringJoin(", "s, previousUnfinishedBlocks)
+	);
+      }
+    }
+    return ResultType::Success();
+  }
+  
+  ResultType VariablesEnvironmentImmutable::getCommonVariables(std::vector<std::string> previousBlocks, std::set<std::string>& commonVariables)
+  {
+    std::vector<std::string> previousFinishedBlocks;
+    std::vector<std::string> previousUnfinishedBlocks;
+
+    for (auto& previousBlock: previousBlocks)
+    {
+      RefinementUtils::Contains(finishedBlocks, previousBlock) ?
+	previousFinishedBlocks.push_back(previousBlock) :
+	previousUnfinishedBlocks.push_back(previousBlock);
+    }
+
+    // In a CFG, it is possible that previous blocks aren't _all_ processed (loops)
+    // However, for immutable constructs, this shouldn't happen. Raise an Error, if it does.
+    if (previousUnfinishedBlocks.size() > 0)
+    {
+      return ResultType::Error("There exist: "s + std::to_string(previousUnfinishedBlocks.size()) + " unfinished Block that were appended without processing. This is an error in an immutable environment."s);
+    }
+
+    bool first = true;
+
+    for (auto& previousBlock : previousBlocks)
+    {
+      if (first)
```

Comment:
Before accessing variablesMappingsPerBlock[previousBlock] there should be a check to make sure variablesMappingsPerBlock[previousBlock] is actually filled in... else the order of iteration is broken

Reply:
See finishedBlocks